internal/app/go-filecoin/internal/submodule/storage_protocol_submodule.go

```diff
@@ -2,22 +2,105 @@ package submodule
+	iface "github.com/filecoin-project/go-fil-markets/storagemarket"
+	impl "github.com/filecoin-project/go-fil-markets/storagemarket/impl"
+	"github.com/ipfs/go-datastore"
+	graphsync "github.com/ipfs/go-graphsync/impl"
+	"github.com/ipfs/go-graphsync/ipldbridge"
+	gsnet "github.com/ipfs/go-graphsync/network"
+	gsstoreutil "github.com/ipfs/go-graphsync/storeutil"
+	"github.com/ipfs/go-hamt-ipld"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/pkg/errors"
+
+	"github.com/filecoin-project/go-filecoin/internal/app/go-filecoin/paths"
+	"github.com/filecoin-project/go-filecoin/internal/app/go-filecoin/plumbing/msg"
+	storagemarketconnector "github.com/filecoin-project/go-filecoin/internal/app/go-filecoin/storage_market_connector"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/piecemanager"
+	fcaddr "github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/wallet"
 )
 
-// StorageProtocolSubmodule enhances the `Node` with "Storage" protocol capabilities.
+// StorageProtocolSubmodule enhances the node with storage protocol
+// capabilities.
 type StorageProtocolSubmodule struct {
-	StorageAPI *storage.API
-
-	// Storage Market Interfaces
-	StorageMiner *storage.Miner
+	StorageClient   iface.StorageClient
+	StorageProvider iface.StorageProvider
 }
 
 // NewStorageProtocolSubmodule creates a new storage protocol submodule.
-func NewStorageProtocolSubmodule(ctx context.Context) (StorageProtocolSubmodule, error) {
-	return StorageProtocolSubmodule{
-		// StorageAPI: nil,
-		// StorageMiner: nil,
+func NewStorageProtocolSubmodule(
+	ctx context.Context,
+	minerAddr fcaddr.Address,
+	clientAddr fcaddr.Address,
+	c *ChainSubmodule,
+	m *MessagingSubmodule,
+	mw *msg.Waiter,
+	pm piecemanager.PieceManager,
+	wlt *wallet.Wallet,
+	h host.Host,
+	ds datastore.Batching,
+	bs blockstore.Blockstore,
+	repoPath string,
+	wg storagemarketconnector.WorkerGetter) (*StorageProtocolSubmodule, error) {
+
+	ma, err := address.NewFromBytes(minerAddr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	ca, err := address.NewFromBytes(clientAddr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	pnode := storagemarketconnector.NewStorageProviderNodeConnector(ma, c.State, m.Outbox, mw, pm, wg, wlt)
+	cnode := storagemarketconnector.NewStorageClientNodeConnector(hamt.CSTFromBstore(bs), c.State, mw, wlt, m.Outbox, ca, wg)
+
+	pieceStagingPath, err := paths.PieceStagingDir(repoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure pieces directory exists
+	err = os.MkdirAll(pieceStagingPath, 0700)
+	if err != nil {
+		return nil, err
+	}
+
+	fs, err := filestore.NewLocalFileStore(filestore.OsPath(pieceStagingPath))
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure miner-address is in datastore because graphsyncimpl requires it
+	err = ds.Put(datastore.NewKey("miner-address"), minerAddr.Bytes())
```

Comment:
This seems quite bad.  If this is a necessary parameter for running a graphsync impl then why not pass this as a parameter to `NewGraphSyncDataTransfer`? 

At the very least the piece datastore should be separated to avoid overwrite deletion by other parts of the application which have no business knowing about this global requirement.

Reply:
agreed. I'll see if I can get it changed in go-fil-markets. It's probably just a convenience coming from lotus.

Reply:
https://github.com/filecoin-project/go-fil-markets/pull/108